8223373: Remove IntelliJ IDEA specific files from the source code repository

tsayao opened the pull request:
This PR does:

- Remove specific Idea files and let it be imported from gradle;
- Adds checkstyle (to use with checkstyle plugin - it will let you know style mistakes);
- Configures auto-format to sun style (with the changes mentioned in [Code Style Rules](https://wiki.openjdk.org/display/OpenJFX/Code+Style+Rules));
- Automatically sets Copyright notice (updates year too);

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8223373](https://bugs.openjdk.org/browse/JDK-8223373): Remove IntelliJ IDEA specific files from the source code repository


### Reviewers
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1009/head:pull/1009` \
`$ git checkout pull/1009`

Update a local copy of the PR: \
`$ git checkout pull/1009` \
`$ git pull https://git.openjdk.org/jfx.git pull/1009/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1009`

View PR using the GUI difftool: \
`$ git pr show -t 1009`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1009.diff">https://git.openjdk.org/jfx/pull/1009.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1009#issuecomment-1407752304)

tsayao commented:
I would appreciate feedback from intellij idea users.

tsayao commented:
CheckStyle plugin is also available for [Eclipse](https://checkstyle.org/eclipse-cs/#!/) if anyone is willing to set it up.

It caught many simple mistakes on the code.

@andy-goryachev-oracle You might like it :)

andy-goryachev-oracle commented:
will take a look at the CheckStyle plug-in, thanks!

andy-goryachev-oracle commented:
Since you are touching /.gitignore, could you also

- Add hs_err*.log, .DS_Store to .gitignore

from JDK-8296810 ?

tsayao commented:
I would like to add manual tests to gradle so it would be imported by the IDE in a way it doesn't build (most are single java files that can be run directly) but code completion works. Does anyone know how to do that?

Maran23 commented:
Since I also use IntelliJ, I will review the PR in the next few days. :)

tsayao commented:
I still cannot run manual tests directly from the IDE. I think it's close to work, but something is missing and I'm not a gradle ninja.

tsayao commented:
Eclipse outputs the same error when running manual tests, so I think it's not possible to fix without restructuring the folder. 

Ready for review.

kevinrushforth commented:
This touches `build.gradle` and `settings.gradle` in addition to IDE-specific files. Those changes will need additional review, so I'm requesting two reviewers.

/reviewers 2

Maran23 commented on .idea/runConfigurations/Ensemble8_jar.xml:
```diff
@@ -0,0 +1,8 @@
+<component name="ProjectRunConfigurationManager">
```
Do we really need these run configurations (except `Run_Robot_Tests.xml`, which is useful)

tsayao replied:
Yeah, I agree. Removed those and left the robot tests.

Maran23 commented on .idea/misc.xml:
```diff
@@ -1,8 +1,11 @@
+  </component>
   <component name="IdProvider" IDEtalkID="9BC0DE4049720C1246C2D3B835C6F7C3" />
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" project-jdk-name="11" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" project-jdk-name="19" project-jdk-type="JavaSDK">
```
I think this should be `project-jdk-name="17"`. When I check out this PR, IntelliJ also changed that automatically.

tsayao replied:
Each user probably would have to configure a JDK for the project.

Maran23 replied:
Yes I agree. But since the language level is JDK-17 and this file is part of the version control, I would recommend we synchronize the `project-jdk-name` with the `languageLevel`, in this case `17`.

tsayao replied:
I've changed it to "17".

Maran23 commented on .idea/runConfigurations/Run_Robot_Tests.xml:
```diff
@@ -0,0 +1,23 @@
+      <option name="executionName" />
+      <option name="externalProjectPath" value="$PROJECT_DIR$" />
+      <option name="externalSystemIdString" value="GRADLE" />
+      <option name="scriptParameters" value="-PFULL_TEST=true -PUSE_ROBOT=true" />
```
Is this also possible for manual tests? Or how do you launch manual tests?

tsayao replied:
I do it by clicking on the run icon of the main method. Manual tests uses non-standard folder structure, so I had to customize the gradle script. As they are not part of the build, it's not possible (I think) to create run configurations.

Maran23 commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,41 @@
+<component name="ProjectCodeStyleConfiguration">
+  <code_scheme name="Project" version="173">
```
We should consider adding `<option name="LINE_SEPARATOR" value="&#10;" />` -> \n as separator

tsayao replied:
Added.

Maran23 commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,41 @@
+<component name="ProjectCodeStyleConfiguration">
+  <code_scheme name="Project" version="173">
```
Maybe `OpenJFX` is a better name here?

tsayao replied:
I tried to change it but didn't work. It seems IntelliJ expects "Project" there.

Maran23 reviewed (commented):
Looks good to me. I just tested it with the newest `IntelliJ` version (2022.3.2) and it works fine.
It also follows what IntelliJ is suggesting. 
Also the gradle changes looks fine to me, although not sure how I can start the manual tests.
I left some comments.

Maran23 reviewed (approved):
Looks good to me. Tested with the latest IntelliJ version, everything works as expected.
The addition to the `.gitignore` is also very helpful and I can confirm I get no unrelated IntelliJ/Sonarlint files anymore.
Gradle changes also looks fine as far as I can tell. But someone else should probably have a look as well.

tsayao commented:
Intellij Ideas does not see the dependencies:
```
dependencies {
        testImplementation project(":graphics").sourceSets.test.output
        testImplementation project(":base").sourceSets.test.output
        testImplementation project(":controls").sourceSets.test.output
        testImplementation project(":swing").sourceSets.test.output
    }
```

How do I rewrite it?

Maran23 commented:
> Intellij Ideas does not see the dependencies:
> 
> ```
> dependencies {
>         testImplementation project(":graphics").sourceSets.test.output
>         testImplementation project(":base").sourceSets.test.output
>         testImplementation project(":controls").sourceSets.test.output
>         testImplementation project(":swing").sourceSets.test.output
> }
> ```
> 
> How do I rewrite it?
> 
> It works if i remove `.sourceSets.test.output`, so I thought maybe It can be rewritten so it's good for Intellij and the build.

I do not know either. But with this PR the setup works already better than before and finally no more files inside my git staging.
So if you agree we can also just merge the IntelliJ changes now and try to fix the remaining issue(s) later, since this PR already improves the situation a lot. :)

tsayao commented:
I agree. I have removed the gradle changed so it's easier to review.

Maran23 reviewed (approved):
LGTM, tested successfully on my Windows machine with IntelliJ.

tsayao commented:
@kevinrushforth I've reverted the gradle changes, so it pretty much only affect the IDE.

kevinrushforth commented on .gitignore:
```diff
@@ -16,13 +16,55 @@ testbin/
+.idea/modules.xml
+.idea/*.iml
+.idea/modules
+*.iml
+*.ipr
+
+# File-based project format
+*.iws
+
+# IntelliJ
+out/
```
This would pick up more than just IntelliJ files. Are they all needed?

Maran23 replied:
`out` is the default `Compiler output` path in IntelliJ. But I don't think it is really needed.

tsayao replied:
Updated it, `*.iml` and `*.ipr` are the Idea project files that are created when importing from gradle.

kevinrushforth commented on .gitignore:
```diff
@@ -16,13 +16,55 @@ testbin/
+.idea/sonarlint/
+#IntelliJ Idea End
+
+#Crash log
+hs_err*.log,
+
+#mac folder
+.DS_Store
```
These seem unrelated to this PR (and in the case of `hs_err`, there is a typo).

tsayao replied:
Removed them

kevinrushforth commented on checkstyle.xml:
```diff
@@ -0,0 +1,213 @@
+<?xml version="1.0"?>
```
I'm not keen on adding yet another top-level file in our repo. Can it go somewhere else?

tsayao replied:
I've looked for a place, but didn't found one that would make sense. It could be used on eclipse as well and probably on netbeans.

tsayao replied:
@kevinrushforth can you guide me what should I do about this file?

kevinrushforth replied:
Since we already have a few other build and IDE-related files in the top repo, and since it could be shared by other IDEs, I won't object to putting it in the top repo.

kevinrushforth reviewed (commented):
> I've reverted the gradle changes, so it pretty much only affect the IDE.

Good. I'll let the IntelliJ users do the review. My only comments are on the two files outside the `.idea` directory.

tsayao commented:
@arapte Would you take a look when you have time? Thanks.

tsayao commented:
> I am testing with IntelliJ IDEA 2023.1.2 (Community Edition). Made some changes in a java file, added empty line with spaces, trailing spaces, added tabs at random places. All the spaces on empty line, trailing spaces were correctly removed, tabs were replaced with spaces by the IDE with or **without** this change. Am I missing anything here? Copyright year was not updated in the modified file. Are there any other steps to follow ? Also, without this change, the gradle tries to build the project as soon as we open the project in IDE, but with this change it does not. which seems OK to me.
> 
> Also I do not use IDE to build jfx nor to execute the tests. If you do, can you please share me any specific steps.

I configured it to follow Sun style with JavaFX exceptions (like allow longer line with). Converting tabs to space and removing trailing spaces are probably default.

About updating copyright year, maybe it's an Ultimate feature that might not be available in Community.

Tests are not working yet, because Idea is not able to correctly import from Gradle without changing it (I tried it but reverted to keep it unchanged for this PR).

arapte reviewed (approved):
Looks good.
Did not face any problem when loading the project in idea, changing properties, editing files.

tsayao commented:
/integrate
